doc/Neighbor/Neighbor_Refresh_HLD.md

```diff
@@ -0,0 +1,321 @@
+
+Req-3:
+
+MAC table entry in SONiC can be administratively cleared or could be removed due to control protocol handling (mclag mac delete on orphan ports during mclag peer reboot). Once the MAC is removed by external triggers then traffic could be dropped up untill the next neighbor refresh which could be in minutes. In order to avoid such traffic drop, if a MAC is removed, neighbor entry should be refreshed as early as possible to avoid blackholing of traffic .
+
+# 3 Design
+## 3.1 Overview
+
+In order to refresh the neighbors, application needs to be aware of all the neighbors in use. Neighsyncd gets the kernel event for every neighbor added in kernel neighbor table, however managing the neighbor refresh in neighbsyncd would mean some of the processing time will be spent on neighbor refresh activies instead of updating NEIGH_TABLE in APP_DB. Hence it is proposed to handle neighbor refresh in another process, since nbrmgrd already handles static neighbor programming and neighbor resolve table (to trigger neighbor resolution in kernel for unresolved neighbor) and to keep all neighbor resolution/refresh to one component, nbrmgrd is selected as the process to handle neighbor refresh processing.
+
+A new NEIGH_REFRESH_TABLE will be added as part of this feature and neighbsyncd in addition to NEIGH_TABLE, will add neighbor entries to NEIGH_REFRESH_TABLE along with additional info like the kernel state and flags.
+
+Nbrmgrd will subscribe to NEIGH_REFRESH_TABLE and create a local cache depending upon the kernel state and flags.
+
+Nbrmgrd will also have a peridic neighbor timer to handle both the Neighbor refresh and the FDB refresh. If a neighbor's refresh timer expires then ARP request or Neighbor solicitation message will be sent using a RAW socket. Nbrmgrd will not wait for the response because waiting for a response will delay sending of subsequent refresh packets and in scale scenario will impact refresh process. Once the packet is transmitted out successfully the response from the destination host will be processed by kernel and kernel will update its internal timers.
+
+## 3.2 DB Changes
+
+### 3.2.1 CONFIG DB
+Not applicable
+
+### 3.2.2 APP DB
+**NEIGH_REFRESH_TABLE**
+
+```
+Producer: neighsyncd 
+
+Consumer: nbrmgrd
+
+Description: Create a neighbor refresh table entry for every neighbor that is added to NEIGH_TABLE 
+```
+
+Schema:
+
+```
+; New table
+; Has the ARP/ND entries that needs to be refreshed periodically
+
+key           = NEIGH_REFRESH_TABLE:ifname:IPprefix
+
+; field       = value
+neigh         = 12HEXDIG           ; mac address of the neighbor
+family        = "IPv4" / "IPv6"    ; address family
+state         = 1*10DIGIT          ; Neighbor state in Kernel
+flags         = 1*10DIGIT          ; Neighbor flag info in Kernel
+
+Example:
+127.0.0.1:63792> hgetall "NEIGH_REFRESH_TABLE:Ethernet25:20.20.20.1"
+ 1) "neigh"
+ 2) "80:a2:35:26:0c:61"
+ 3) "family"
+ 4) "IPv4"
+ 5) "state"
+ 6) "2"
```

Comment:
Why cant we have the meaningful state names instead of numbers?

Reply:
The state number directly map with kernel states like NUD_REACHABLE, else there will be string validation rather than int. If for any reason if the state has bitwise ORed values it will be directly validated in corresponding subscriber modules.